core-chart/templates/argocd/operator.yaml

```diff
@@ -0,0 +1,163 @@
+
+    timestamped_echo "Logging into Argo CD cli..."
+    while ! kubectl -n $NS exec $ARGOCD_POD -- argocd login "{{ .Release.Name }}-argocd-server.$NS" --username admin --password $(echo $ARGOCD_PASSWD | base64 -d) --insecure ; do
+      >&2 timestamped_echo "==> ERROR: Argo CD login failed! ...retrying in $WAIT_BEFORE_RETRY seconds..."
+      sleep $WAIT_BEFORE_RETRY
+    done
+    timestamped_echo "==> Argo CD login succesful."
+
+    timestamped_echo "Adding WDS clusters..."
+    {{- range $cp := .Values.WDSes }}
+
+      cp_name={{ $cp.name }}
+
+      timestamped_echo "==> Looking for Control Plane \"$cp_name\"..."
+      # Wait for the CP to exists and be ready
+      while [[ $(kubectl get controlplane $cp_name -o 'jsonpath={..status.conditions[?(@.type=="Ready")].status}') != "True" ]]; do
+        >&2 timestamped_echo "==> ERROR: Control Plane \"$cp_name\" is not ready! ...retrying in $WAIT_BEFORE_RETRY seconds..."
+        sleep $WAIT_BEFORE_RETRY
+      done
+      timestamped_echo "==> Control Plane \"$cp_name\" is ready."
+
+      # determine the secret name and namespace
+      key=$(kubectl get controlplane $cp_name -o=jsonpath='{.status.secretRef.inClusterKey}')
+      secret_name=$(kubectl get controlplane $cp_name -o=jsonpath='{.status.secretRef.name}')
+      secret_namespace=$(kubectl get controlplane $cp_name -o=jsonpath='{.status.secretRef.namespace}')
+
+      kubeconfig="/tmp/${cp_name}.kubeconfig"
+
+      timestamped_echo "==> Getting \"$key\" from \"$secret_name\" secret in \"$secret_namespace\" for control plane \"$cp_name\"... \"$kubeconfig\""
+      kubectl get secret $secret_name -n $secret_namespace -o=jsonpath="{.data.$key}" | base64 -d > "$kubeconfig"
+
+      timestamped_echo "==> Copying kubeconfig \"${cp_name}.kubeconfig\" to Argo CD pod \"$ARGOCD_POD\"..."
+      kubectl -n $NS cp "$kubeconfig" ${ARGOCD_POD}:/tmp
+      rm "$kubeconfig"
+
+      timestamped_echo "==> Registering Control Plane \"$cp_name\" as an Argo CD cluster..."
+      kubectl -n $NS exec $ARGOCD_POD -- argocd cluster add ${cp_name} -y --kubeconfig "$kubeconfig"
+      kubectl -n $NS exec $ARGOCD_POD -- rm "$kubeconfig"
+
+    {{- end }}
+
+    timestamped_echo "Creating applications..."
+    {{- range $app := (index .Values "argo-cd" "applications") }}
+
+      timestamped_echo "==> Creating application \"{{ $app.name }}\":"
+      timestamped_echo "    - project=\"{{ $app.project | default "default" }}\""
+      timestamped_echo "    - repoURL=\"{{ $app.repoURL }}\""
+      timestamped_echo "    - targetRevision=\"{{ $app.targetRevision | default "HEAD" }}\""
+      timestamped_echo "    - path=\"{{ $app.path | default "/" }}\""
+      timestamped_echo "    - destinationWDS=\"{{ $app.destinationWDS }}\""
+      timestamped_echo "    - destinationNamespace=\"{{ $app.destinationNamespace | default "default" }}\""
+
+      kubectl exec $ARGOCD_POD -- argocd app create {{ $app.name }} \

```

Comment:
All these user-provided string values should be quoted to prevent shell injection attacks (equivalently, be robust agains mistakes that inject characters interpreted by the shell). Unfortunately, I am not familiar with a function available to Helm templates that does exactly quoting/escaping as needed for bash. The closest that I can recommend is `js`, which formats the string as in JavaScript.

Reply:
I see that you put literal dquotes in the source. That is a good start. We should do better in the future.